Read config from pyproject.toml

edgarrmondragon opened the pull request:
This PR adds support for `pyproject.toml` as a configuration source.

It would help keeping all configuration for a Python project in a single file, since this file is used by other major Python projects like [black] and [pytest].

Some implementation details:

- Takes precedence over `.sqlfluff`.

- `core` section lives in `[tool.sqlfluff.core]`.

    ```toml
    [tool.sqlfluff.core]
    templater = "jinja"
    sql_file_exts = [
        ".sql",
        ".sql.j2",
        ".dml",
        ".ddl",
    ]

    [tool.sqlfluff.indentation]
    indented_joins = false
    template_blocks_indent = false

    [tool.sqlfluff.templater]
    unwrap_wrapped_queries = true

    [tool.sqlfluff.templater.jinja]
    apply_dbt_builtins = true
    ```

Closes #890

[black]: https://black.readthedocs.io/en/stable/pyproject_toml.html
[pytest]: https://docs.pytest.org/en/6.2.x/customize.html#pyproject-toml

barrywhart commented on docs/source/configuration.rst:
```diff
@@ -19,13 +19,44 @@ any vales read from earlier files.
+*[sqlfluff:jinjacontext]*.
+
+For the `pyproject.toml file`_, all valid sections start with `tool.sqlfluff`
+and subsections are delimited by a dot and suffixed with *__options*.
```
You mentioned the `__options` suffix in the PR description. Can you give an example of the kind of conflict you're avoiding by doing this?

If I understand correctly, could the toml reader skip sections that don't start with `tools.sqlfluff` or `sqlfluff`?

edgarrmondragon replied:
So a TOML is basically a nested  object/dictionary so the keys would collide. The following code raises an error

```python
import toml

content = """
[tool.sqlfluff]
templater = "jinja"

[tool.sqlfluff.templater]
unwrap_wrapped_queries = true
"""

toml.loads(content)
```

The error is:

```
What? templater already exists?{'templater': 'jinja'} (line 5 column 1 char 38)
```

> If I understand correctly, could the toml reader skip sections that don't start with tools.sqlfluff or sqlfluff?

Yes, the other sections are ignored. The convention I believe is to put every set of configurations for `my_tool` under `[tool.my_tool*]` sections.

barrywhart commented on src/sqlfluff/core/config.py:
```diff
@@ -144,6 +144,34 @@ def get_global(cls) -> "ConfigLoader":
+        The return value is a list of tuples, were each tuple has two elements,
+        the first is a tuple of paths, the second is the value at that path.
+        """
+        import toml
```
Could this import be moved to the top of the file? That's generally the preferred location for imports.

edgarrmondragon replied:
Sure. I'll do that.

barrywhart reviewed (changes requested):
The PR looks pretty good. I would like to understand the reason for the "__options" suffix in the .toml file. I wonder if there may be other options.

Also, I think we need to update the configuration docs in `ocs/source/configuration.rst`, which documents how configuration works, what files it looks at, etc.

edgarrmondragon commented:
@barrywhart 

> The PR looks pretty good. I would like to understand the reason for the "__options" suffix in the .toml file. I wonder if there may be other options.

I think one option, though it would existing sections names, would be to avoid the same keys for settings and sections, e.g. `templater=jinja` and `[sqlfluff:templater]`.

> Also, I think we need to update the configuration docs in `ocs/source/configuration.rst`, which documents how configuration works, what files it looks at, etc.

I added a few lines and an example TOML block to the first part of the configuration page. Also added `pyproject.toml` to the list of files. Do you think I'm missing something else?

barrywhart commented on src/sqlfluff/core/config.py:
```diff
@@ -144,6 +146,31 @@ def get_global(cls) -> "ConfigLoader":
+        the first is a tuple of paths, the second is the value at that path.
+        """
+        config = toml.load(fpath)
+        tool = config.get("tool", {}).get("sqlfluff", {})
```
Now that you replaced `__options` with the `tool.sqlfluff.core` technique, do you need to move or place the `core` values to match the structure of the other config file types?

In other words, have you verified that SQLFluff correctly uses the values from the `core` section?
```
[tool.sqlfluff.core]
    templater = "jinja"
    sql_file_exts = [
        ".sql",
        ".sql.j2",
        ".dml",
        ".ddl",
    ]
```

edgarrmondragon replied:
@barrywhart I added a test case with

```toml
[tool.sqlfluff.core]
testing_int = 5
testing_bar = 7.698
testing_bool = false
testing_arr = [ "a", "b", "c" ]
testing_inline_table = { x = 1 }
 
[tool.sqlfluff.bar]
foo = "foobar"

[tool.sqlfluff.fnarr.fnarr]
foo = "foobar"
```

which gets collected into the following dict:

```python
{
    "core": {
        "testing_int": 5,
        "testing_bar": 7.698,
        "testing_bool": False,
        "testing_arr": ["a", "b", "c"],
        "testing_inline_table": {"x": 1},
    },
    "bar": {"foo": "foobar"},
    "fnarr": {"fnarr": {"foo": "foobar"}},
}
```

does it look right?``

barrywhart replied:
TBH, I'd feel much better if you could do some manual tests to confirm the settings in the TOML file are being read correctly and actually affecting the runtime behavior of SQLFluff. There are sometimes limits to the effectiveness of automated tests and manual review, and I think this is one of those times.

edgarrmondragon replied:
@barrywhart  That makes a lot of sense. I created [this repo](https://github.com/edgarrmondragon/sqlfluff-pyproject) to test some configurations. Let me know if you think there are edge cases I could add.

barrywhart replied:
I tested with your example repo. Works great! I also added a simple test with templating. This is based on the Variable Templating [example](https://docs.sqlfluff.com/en/stable/configuration.html#variable-templating) from the SQLFluff docs.

`test.sql`:
```
SELECT {{ num_things }} FROM {{ tbl_name }} WHERE id > 10 LIMIT 5
```

Additions to `pyproject.toml`:
```
[tool.sqlfluff.core]
templater = "jinja"

[tool.sqlfluff.templater.jinja.context]
num_things=456
tbl_name="my_table"
```

edgarrmondragon replied:
I added that templated example to the example repo :smile:.

barrywhart reviewed (approved):
Great work! Thanks for adding the test project. 👍
